IPython/html/static/widgets/js/widget.js

```diff
@@ -307,68 +350,70 @@ define(["widgets/js/manager",
             this._buffered_state_diff = {};
         },
 
-        save_changes: function(callbacks) {
-            /**
-             * Push this model's state to the back-end
-             *
-             * This invokes a Backbone.Sync.
-             */
-            this.save(this._buffered_state_diff, {patch: true, callbacks: callbacks});
-        },
 
-        _pack_models: function(value) {
-            /**
-             * Replace models with model ids recursively.
-             */
+        send_sync_message: function(attrs, callbacks) {
+            // prepare and send a comm message syncing attrs
+            var value;
             var that = this;
-            var packed;
-            if (value instanceof Backbone.Model) {
-                return "IPY_MODEL_" + value.id;
-
-            } else if ($.isArray(value)) {
-                packed = [];
-                _.each(value, function(sub_value, key) {
-                    packed.push(that._pack_models(sub_value));
-                });
-                return packed;
-            } else if (value instanceof Date || value instanceof String) {
-                return value;
-            } else if (value instanceof Object) {
-                packed = {};
-                _.each(value, function(sub_value, key) {
-                    packed[key] = that._pack_models(sub_value);
-                });
-                return packed;
-
-            } else {
-                return value;
+            // first, serialize all attrs values, using the serializer if it exists
+            var newattr = _.reduce(attrs, function(obj, v, k) {
+                if (that.serialize[k]) {
+                    return that.serialize[k].then(function(f) {
+                        if (f.serialize) {
+                            obj[k] = f.serialize(v);
+                        } else {
+                            obj[k] = v;
+                        }
+                    })
+                } else {
+                    obj[k] = v;
+                }
+                return obj
+            }, {});
+
+            utils.resolve_promises_dict(newattr).then(function(state) {
+                // get binary values, then send
+                var keys = Object.keys(state);
+                var buffers = [];
+                var buffer_keys = [];
+                for (var i=0; i<keys.length; i++) {
+                    var key = keys[i];
+                    var value = state[key];
+                    if (value.buffer instanceof ArrayBuffer) {
+                        value = value.buffer;
+                    }
```

Comment:
I don't think I understand the if statement above.  In case I do, is setting `.buffer` of a value a convention in Javascript?  If not, why would we want to support that?  It seems like value itself should be an ArrayBuffer if someone wants to use the binary msg spec.